### PR TITLE
refactor ZktrieState [dup #903]

### PR DESCRIPTION
**Pull request by noel2004**
This PR duplicated the target of #903 with another approach:

We consider:

1. A `CircuitInputBuilder` object should not be reused for building more witness block after it is used inside `block_traces_to_witness_block_with_updated_state`. as the result, we can updated the `ZktrieState` memeber in it into the zktrie state after the witness block is applied, just like the `state/code_db` members.
2. We also notice that `Witnessgenerator` do not need a snapshot about account /storage states since all the information can be read from the underlying database (`ZkMemoryDb` here) with trivial cost. So `ZktrieState` also do not need to keep such a snapshot inside it (they have been kept in state db).
3. With refactoring above, and thanks to the lambda nature of trie's storage. `ZktrieState` now can esaily switch to any state it have been once updated to, and create new `ZkTrie` object with almost no cost, which is ready for being updated by another witness block.

### CODE REVIEW

File: zktrie/Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ license = "MIT OR Apache-2.0"
 [dependencies]
 halo2_proofs = { git = "https://github.com/privacy-scaling-explorations/halo2.git", tag = "v2023_02_02" }
 mpt-circuits = { package = "halo2-mpt-circuits", git = "https://github.com/scroll-tech/mpt-circuit.git", tag = "v0.7.0" }
-zktrie = { git = "https://github.com/scroll-tech/zktrie.git", branch = "v0.6" }
+zktrie = { git = "https://github.com/scroll-tech/zktrie.git" }
```
Comment (lispc) [inline zktrie/Cargo.toml]: do we need a new branch?

**Review by lispc — APPROVED** [review]
nothing very bad but i cannot review line by line.  can be merged first.

File: zktrie/Cargo.toml (hunk 1 above)
Comment (noel2004) [reply zktrie/Cargo.toml]: Yes, after test we would push `v0.7` for the new zktrie

**Comment by noel2004** [comment]
Have run mock proving on 40 batches (>200 chunks) from replayed alpha testnet, so far so good

**Pull request merged**